tools/make_patch.c

```diff
@@ -0,0 +1,944 @@
+		if (!file) {
+			fprintf(stderr, "Error: Cannot open file: %s\n", filename);
+			goto error;
+		}
+		symbols = parse_asm(file);
+		fclose(file);
+		file = NULL;
+
+		// Save them for later re-use
+		asmfile = malloc(sizeof(struct asmfile) + strlen(filename) + 1);
+		if (!asmfile) {
+			fprintf(stderr, "Error: Cannot allocate memory\n");
+			if (symbols) free_list(symbols);
+			goto error;
+		}
+		asmfile->symbols = symbols;
+		strcpy(asmfile->name, filename);
+		asmfile->next = asmfiles;
+		asmfiles = asmfile;
+	}
+
+	// Look for the symbol
+	symbol = find_symbol(symbols, constant);
+	if (!symbol) {
+		fprintf(stderr, "Error: Could not find constant %s in file %s\n", constant, filename);
+		goto error;
+	}
+
+	value = symbol->value;
+
+error:
+	if (file) fclose(file);
+	return value;
+}
+
+void interpret_command(char *command, const struct symbol *currentpatch, const struct symbol *symbols, 
+                       struct patch *patch, FILE *new_rom, FILE *orig_rom, FILE *output)
+{
+	int argc = 0;
+	int offset = -1;
+	if (currentpatch) {
+		offset = currentpatch->value;
+	}
+	char *s;
+
+	const struct symbol *getsymbol = NULL;
+	// Count the arguments
+	s = command;
+	while (*s) if (*s++ == ':') argc++;
+
+	// Get the arguments
+	char *argv[argc];
+	if ((s = strchr(command, ':'))) {
+		*s++ = '\0';
+		argv[0] = s;
+		for (int i = 1; i < argc; i++) {
+			if (!(s = strchr(s, ':'))) break;
+			*s++ = '\0';
+			argv[i] = s;
+		}
+	}
+
+	// Now we finally can do stuff with it
+	if (strcmp(command, "ADDREss") == 0) {
+		// Shitty hack, shouldn't be used anyway
+		int high = offset >> 8;
+		int low = offset & 0xFF;
+		fprintf(output, "0x%X%x", high, low);
+	} else if (strcmp(command, "Address") == 0 ||
+			   strcmp(command, "address") == 0) {
+		if (argc > 0) offset += strtol(argv[0], NULL, 0);
+
+		fprintf(output, "0x");
+		if (argc > 1) {
+			for (int i = strtol(argv[1], NULL, 0); i > 0; i--) {
+				putchar('0');
+			}
+		}
+
+		fprintf(output, isupper(command[0]) ? "%X" : "%x", offset);
+	} else if (strcmp(command, "Patch") == 0 ||
+			   strcmp(command, "patch") == 0) {
+		if (argc > 0) offset += strtol(argv[0], NULL, 0);
+
+		if (fseek(orig_rom, offset, SEEK_SET) != 0) {
+			fprintf(stderr, "Error: Could not seek to the offset of %s in the original ROM\n", currentpatch->name);
+			return;
+		}
+
+		if (fseek(new_rom, offset, SEEK_SET) != 0) {
+			fprintf(stderr, "Error: Could not seek to offset of %s in the new ROM\n", currentpatch->name);
+			return;
+		}
+
+		if (argc <= 1 || strcmp(argv[1], "big") != 0) {
+			int c = getc(new_rom);
+			if (c == getc(orig_rom)) {
+				fprintf(stderr, "Warning: %s doesn't actually contain any differences\n", currentpatch->name);
+			}
+
+			patch->offset = offset;
+			patch->size = 1;
+			fprintf(output, "0x");
+			fprintf(output, isupper(command[0]) ? "%02X" : "%02x", c);
+		} else {
+			int length = 0;
+
+			// Figure out the length of the patch
+			char *searchend = malloc(strlen(currentpatch->name) + 5);
```

Comment:
Here and elsewhere, keep an eye out for arbitrary hard-coded numbers. See if they can be expressed in terms of more meaningful values.
```suggestion
			char *searchend = malloc(strlen(currentpatch->name) + strlen("_End") + 1);
```

Reply:
@Rangi42 I defined the addresstype locations in get_address_type_limit. If you don't like it. Feel free to revert the commit.

See https://github.com/pret/pokecrystal/pull/882/commits/7aceaed7b8f11acb016752050e643ce1b37d6bc2